### ISSUE TITLE
Missing function signature when using EM_ASYNC_JS and ASYNCIFY=2

### ISSUE THREAD

**Issue by brendandahl** [emscripten-core/emscripten#17890]
Error message:
`Aborted(Assertion failed: Missing __sig for __asyncjs__promise42)`

Building with:
`emcc -sASSERTIONS -sENVIRONMENT=web -g3 -sASYNCIFY=2 -sEXIT_RUNTIME=1 main.cpp -o a.out.html`

main.cpp
```c
#include <stdio.h>
#include <emscripten.h>

EM_ASYNC_JS(int, promise42, (), {
  return new Promise((k,r)=>{setTimeout(()=> {return k(42)},0)});
});

int main() {
  printf("p1\n");
  promise42();
  printf("p2\n");
  return 0;
}

```

I need to look into why the signatures aren't being generated for EM_ASYNC_JS functions.

**Comment by sbc100** [comment]
I think I ran into this for wasm64 too.

Do we have enough information to generate those signatures?    I'm guessing not... we do have the C signature but its hard to go from that to a `__sig` in a reliable way I think.    Perhaps if we only accepted core C types such as `size_t`, `long`, etc, we could map them.. but if they use any custom types or macros it wouldn't be possible.

**Comment by sbc100** [comment]
Ah yes, here is the same issue but for wasm64: #16975

**Comment by brendandahl** [comment]
Yeah, if we do it only in the preprocessor that would be brittle. I was thinking we could do something like embind and build an array of the mapped types at compile time and extract them in binaryen. However, when I started looking into that I think I found a better way. We already have the signature of the import defined in Wasm, so we could modify the code that extracts the JS functions and have it also extract the signature of the import.
